lib/compiler/src/engine/unwind/systemv.rs

```diff
@@ -138,51 +138,71 @@ impl UnwindRegistry {
-            // scenario. Usually, this is handled upstream by the
-            // compilers.
-            debug_assert_ne!(
-                eh_frame,
-                &[0, 0, 0, 0],
-                "`eh_frame` seems to contain empty FDEs"
-            );
+        #[cfg(all(target_os = "macos", target_arch = "aarch64"))]
+        {
+            // Special call for macOS on aarch64 to register the `.eh_frame` section.
+            // TODO: I am not 100% sure if it's correct to never deregister the `.eh_frame` section. It was this way before
+            // I started working on this, so I kept it that way.
+            compact_unwind::__unw_add_dynamic_eh_frame_section(eh_frame.as_ptr() as usize);
+            return;
+        }
 
-            // On gnu (libgcc), `__register_frame` will walk the FDEs until an entry of length 0
-            let ptr = eh_frame.as_ptr();
-            __register_frame(ptr);
-            self.registrations.push(ptr as usize);
-        } else {
-            #[cfg(all(target_os = "macos", target_arch = "aarch64"))]
-            {
-                compact_unwind::__unw_add_dynamic_eh_frame_section(eh_frame.as_ptr() as usize);
-            }
+        // Validate that the `.eh_frame` is well-formed before registering it.
+        // See https://refspecs.linuxfoundation.org/LSB_3.0.0/LSB-Core-generic/LSB-Core-generic/ehframechpt.html for more details.
+        // We put the frame records into a vector before registering them, because
+        // calling `__register_frame` with invalid data can cause segfaults.
 
-            #[cfg(not(all(target_os = "macos", target_arch = "aarch64")))]
-            {
-                // For libunwind, `__register_frame` takes a pointer to a single FDE
-                let start = eh_frame.as_ptr();
-                let end = start.add(eh_frame.len());
-                let mut current = start;
-
-                // Walk all of the entries in the frame table and register them
-                while current < end {
-                    let len = std::ptr::read::<u32>(current as *const u32) as usize;
-
-                    // Skip over the CIE and zero-length FDEs.
-                    // LLVM's libunwind emits a warning on zero-length FDEs.
-                    if current != start && len != 0 {
-                        __register_frame(current);
-                        self.registrations.push(current as usize);
-                    }
+        // Pointers to the registrations that will be registered with `__register_frame`.
+        // For libgcc based systems, these are CIEs.
+        // For libunwind based systems, these are FDEs.
+        let mut records_to_register = Vec::new();
 
-                    // Move to the next table entry (+4 because the length itself is not inclusive)
-                    current = current.add(len + 4);
+        let mut current = 0;
+        let mut last_len = 0;
+        while current < (eh_frame.len() - 3) {
```

Comment:
Comment please for where the `3` comes from, for the future reader

Reply:
Changed it to `<= len - size_of::<u32>` to make it clear that this is to make sure that it is always safe to cast the current index to an u32